File: easy/phone-number.js
const AREA_CODE_INDEX = 0;
const EXCHANGE_CODE_INDEX = 3;
const VALID_NUMBER_LENGTH = 10;
const MAX_NUMBER_LENGTH = 11;

export const clean = (number) => {
  if (/[a-z]/gi.test(number)) throw new Error("Letters not permitted");
  if (/[!@:]/gi.test(number)) throw new Error("Punctuations not permitted");

  let cleanString = number.match(/\d/g).join("");

  if (cleanString.length < VALID_NUMBER_LENGTH)
    throw new Error("Incorrect number of digits");
  if (cleanString.length > MAX_NUMBER_LENGTH)
    throw new Error("More than 11 digits");
  if (cleanString.length === MAX_NUMBER_LENGTH) {
    if (cleanString[0] !== "1") throw new Error("11 digits must start with 1");
    cleanString = cleanString.slice(1); // if  the country code is included, remove it
  }
  if (cleanString[AREA_CODE_INDEX] === "0")
    throw new Error("Area code cannot start with zero");
  if (cleanString[AREA_CODE_INDEX] === "1")
    throw new Error("Area code cannot start with one");
  if (cleanString[EXCHANGE_CODE_INDEX] === "0")
    throw new Error("Exchange code cannot start with zero");
  if (cleanString[EXCHANGE_CODE_INDEX] === "1")
    throw new Error("Exchange code cannot start with one");

  return cleanString;
};

File: python/easy/phone_number.py
import re

class PhoneNumber:
    def __init__(self, number):
        self.number = self.process_number(number)
        self.area_code = self.number[0:3]

    def process_number(self, number):
        just_digits = self.sanitize_number(number)
        correct_length = self.check_length(just_digits)
        self.check_area_exchange(correct_length)
        return "".join(correct_length)

    def sanitize_number(self, number):
        if re.search("[a-zA-Z]", number):
            raise ValueError("letters not permitted")
        elif re.search("[@!:]", number):
            raise ValueError("punctuations not permitted")
        return re.findall("\d", number)

    def check_length(self, number):
        if len(number) < 10:
            raise ValueError("incorrect number of digits")
        elif len(number) == 11:
            if number[0] != "1":
                raise ValueError("11 digits must start with 1")
            else:
                return number[1:]
        elif len(number) > 11:
            raise ValueError("more than 11 digits")
        else:
            return number

    def check_area_exchange(self, ten_digit_number):
        if ten_digit_number[0] == "0":
            raise ValueError("area code cannot start with zero")
        elif ten_digit_number[0] == "1":
            raise ValueError("area code cannot start with one")
        elif ten_digit_number[3] == "0":
            raise ValueError("exchange code cannot start with zero")
        elif ten_digit_number[3] == "1":
            raise ValueError("exchange code cannot start with one")

    def pretty(self):
        return f"({self.area_code})-{self.number[3:6]}-{self.number[6:]}"